File: craft/commands/vm.py
import click

from craft.client import delete, get, patch, post
from craft.output import print_item, print_json, print_success, print_table


@click.command("list")
@click.option("--page", default=1, help="Page number")
@click.option("--limit", default=10, help="Items per page (max 50)")
def vm_list(page, limit):
    """List your VMs."""
    data = get("/vms", params={"page": page, "limit": limit})
    items = data.get("data", data) if not isinstance(data.get("data"), list) else data["data"]
    if isinstance(items, dict):
        items = items.get("items", items.get("vms", [items]))
    if isinstance(items, list) and items:
        rows = []
        for vm in items:
            rows.append([
                vm.get("id", ""),
                vm.get("name", ""),
                vm.get("status", ""),
                vm.get("ip", vm.get("ipAddress", "")),
                vm.get("os", vm.get("osTemplate", "")),
            ])
        print_table(rows, ["ID", "Name", "Status", "IP", "OS"])
    else:
        print_json(data)


@click.command("get")
@click.argument("vm_id")
def vm_get(vm_id):
    """Get VM details."""
    data = get(f"/vms/{vm_id}")
    print_item(data)


@click.command("create")
@click.option("--name", required=True, help="VM name (1-50 chars)")
@click.option("--hostname", required=True, help="Hostname (RFC-1123)")
@click.option("--node-id", required=True, help="Node UUID")
@click.option("--os-template-id", required=True, help="OS template UUID")
@click.option("--root-password", prompt=True, hide_input=True, help="Root password (min 8 chars)")
@click.option("--plan-id", default=None, help="Plan UUID")
@click.option("--billing-type", type=click.Choice(["daily", "weekly", "monthly", "yearly"]), default="daily")
@click.option("--cpu", default=None, type=int, help="Custom CPU cores (1-32)")
@click.option("--ram-mb", default=None, type=int, help="Custom RAM in MB (512-131072)")
@click.option("--disk-gb", default=None, type=int, help="Custom disk in GB (10-2000)")
@click.option("--ssh-keys", default=None, help="Comma-separated SSH key IDs")
def vm_create(name, hostname, node_id, os_template_id, root_password, plan_id,
              billing_type, cpu, ram_mb, disk_gb, ssh_keys):
    """Create a new VM."""
    body = {
        "name": name,
        "hostname": hostname,
        "nodeId": node_id,
        "osTemplateId": os_template_id,
        "rootPassword": root_password,
        "billingType": billing_type,
    }
    if plan_id:
        body["planId"] = plan_id
    if cpu:
        body["cpu"] = cpu
    if ram_mb:
        body["ramMb"] = ram_mb
    if disk_gb:
        body["diskGb"] = disk_gb
    if ssh_keys:
        body["sshKeys"] = ssh_keys

    data = post("/vms", body)
    print_success("VM created.")
    print_item(data)


@click.command("delete")
@click.argument("vm_id")
@click.confirmation_option(prompt="This will permanently delete the VM and all data. Continue?")
def vm_delete(vm_id):
    """Delete a VM permanently."""
    delete(f"/vms/{vm_id}")
    print_success(f"VM {vm_id} deleted.")


@click.command("status")
@click.argument("vm_id")
def vm_status(vm_id):
    """Get real-time VM status."""
    data = get(f"/vms/{vm_id}/status")
    print_item(data)


@click.command("start")
@click.argument("vm_id")
def vm_start(vm_id):
    """Start a VM."""
    post(f"/vms/{vm_id}/start")
    print_success(f"VM {vm_id} started.")


@click.command("stop")
@click.argument("vm_id")
def vm_stop(vm_id):
    """Stop a VM (graceful ACPI shutdown)."""
    post(f"/vms/{vm_id}/stop")
    print_success(f"VM {vm_id} stopped.")


@click.command("reboot")
@click.argument("vm_id")
def vm_reboot(vm_id):
    """Reboot a VM."""
    post(f"/vms/{vm_id}/reboot")
    print_success(f"VM {vm_id} rebooted.")


@click.command("rename")
@click.argument("vm_id")
@click.option("--name", required=True, help="New name (1-50 chars)")
def vm_rename(vm_id, name):
    """Rename a VM."""
    patch(f"/vms/{vm_id}/rename", {"name": name})
    print_success(f"VM renamed to '{name}'.")


@click.command("reset-password")
@click.argument("vm_id")
@click.option("--username", default="root", help="Username (default: root)")
@click.option("--new-password", prompt=True, hide_input=True, confirmation_prompt=True)
def vm_reset_password(vm_id, username, new_password):
    """Reset VM user password via QEMU Guest Agent."""
    post(f"/vms/{vm_id}/reset-password", {"username": username, "newPassword": new_password})
    print_success(f"Password reset for '{username}'.")


@click.command("reset-network")
@click.argument("vm_id")
def vm_reset_network(vm_id):
    """Re-apply VM network configuration."""
    post(f"/vms/{vm_id}/reset-network")
    print_success("Network reset. Reboot the VM to apply.")


@click.command("resize")
@click.argument("vm_id")
@click.option("--plan-id", required=True, help="Target plan UUID")
def vm_resize(vm_id, plan_id):
    """Resize VM to a new plan."""
    post(f"/vms/{vm_id}/resize", {"planId": plan_id})
    print_success("VM resized.")


@click.command("reinstall")
@click.argument("vm_id")
@click.option("--os-template-id", required=True, help="OS template UUID")
@click.option("--root-password", prompt=True, hide_input=True)
@click.option("--ssh-keys", default=None, help="Comma-separated SSH key IDs")
@click.confirmation_option(prompt="WARNING: All data will be destroyed. Continue?")
def vm_reinstall(vm_id, os_template_id, root_password, ssh_keys):
    """Reinstall OS on a VM."""
    body = {"osTemplateId": os_template_id, "rootPassword": root_password}
    if ssh_keys:
        body["sshKeys"] = ssh_keys
    post(f"/vms/{vm_id}/reinstall", body)
    print_success("OS reinstalled.")


@click.command("console")
@click.argument("vm_id")
def vm_console(vm_id):
    """Get noVNC console access."""
    data = get(f"/vms/{vm_id}/console")
    print_item(data)


@click.command("network")
@click.argument("vm_id")
def vm_network(vm_id):
    """Get VM network info."""
    data = get(f"/vms/{vm_id}/network")
    print_item(data)


@click.command("metrics")
@click.argument("vm_id")
@click.option("--hours", default=24, help="Hours of data (1-168, default: 24)")
def vm_metrics(vm_id, hours):
    """Get VM performance metrics."""
    data = get(f"/vms/{vm_id}/metrics", params={"hours": hours})
    print_json(data)

File: cmd/vm.go
package cmd

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

func init() {
	// vm list
	vmListCmd.Flags().Int("page", 1, "Page number")
	vmListCmd.Flags().Int("limit", 20, "Items per page")
	VMCmd.AddCommand(vmListCmd)

	// vm get
	VMCmd.AddCommand(vmGetCmd)

	// vm create
	vmCreateCmd.Flags().BoolP("interactive", "i", false, "Interactive wizard mode")
	vmCreateCmd.Flags().String("name", "", "VM name")
	vmCreateCmd.Flags().String("hostname", "", "VM hostname")
	vmCreateCmd.Flags().String("node-id", "", "Node ID")
	vmCreateCmd.Flags().String("os-template-id", "", "OS template ID")
	vmCreateCmd.Flags().String("root-password", "", "Root password")
	vmCreateCmd.Flags().String("plan-id", "", "Plan ID")
	vmCreateCmd.Flags().String("billing-type", "", "Billing type (daily/weekly/monthly/yearly)")
	vmCreateCmd.Flags().Int("cpu", 0, "Number of CPUs")
	vmCreateCmd.Flags().Int("ram-mb", 0, "RAM in MB")
	vmCreateCmd.Flags().Int("disk-gb", 0, "Disk size in GB")
	vmCreateCmd.Flags().String("ssh-keys", "", "Comma-separated SSH key IDs")
	VMCmd.AddCommand(vmCreateCmd)

	// vm delete
	VMCmd.AddCommand(vmDeleteCmd)

	// vm status
	VMCmd.AddCommand(vmStatusCmd)

	// vm start
	VMCmd.AddCommand(vmStartCmd)

	// vm stop
	VMCmd.AddCommand(vmStopCmd)

	// vm reboot
	VMCmd.AddCommand(vmRebootCmd)

	// vm rename
	vmRenameCmd.Flags().String("name", "", "New VM name")
	VMCmd.AddCommand(vmRenameCmd)

	// vm reset-password
	vmResetPasswordCmd.Flags().String("username", "", "Username")
	vmResetPasswordCmd.Flags().String("new-password", "", "New password")
	VMCmd.AddCommand(vmResetPasswordCmd)

	// vm reset-network
	VMCmd.AddCommand(vmResetNetworkCmd)

	// vm resize
	vmResizeCmd.Flags().String("plan-id", "", "New plan ID")
	VMCmd.AddCommand(vmResizeCmd)

	// vm reinstall
	vmReinstallCmd.Flags().String("os-template-id", "", "OS template ID")
	vmReinstallCmd.Flags().String("root-password", "", "Root password")
	vmReinstallCmd.Flags().String("ssh-keys", "", "Comma-separated SSH key IDs")
	VMCmd.AddCommand(vmReinstallCmd)

	// vm console
	VMCmd.AddCommand(vmConsoleCmd)

	// vm network
	VMCmd.AddCommand(vmNetworkCmd)

	// vm metrics
	vmMetricsCmd.Flags().Int("hours", 24, "Number of hours of metrics to retrieve")
	VMCmd.AddCommand(vmMetricsCmd)
}

// resolveVMID returns the VM ID from args or prompts the user to select one.
func resolveVMID(args []string) string {
	if len(args) > 0 && args[0] != "" {
		return args[0]
	}
	return SelectVM("Select a VM")
}

// ---------- vm list ----------

var vmListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all virtual machines",
	Run: func(cmd *cobra.Command, args []string) {
		page, _ := cmd.Flags().GetInt("page")
		limit, _ := cmd.Flags().GetInt("limit")

		params := map[string]string{
			"page":  strconv.Itoa(page),
			"limit": strconv.Itoa(limit),
		}

		result := client.Get("/vms", params, true)
		vms := extractList(result, "vms", "items")

		headers := []string{"ID", "Name", "Status", "IP", "OS"}
		rows := make([][]string, len(vms))
		for i, vm := range vms {
			ip := getString(vm, "ip")
			if ip == "" {
				ip = getString(vm, "ipAddress")
			}
			osName := getString(vm, "os")
			if osName == "" {
				osName = getString(vm, "os_template")
			}
			rows[i] = []string{
				getString(vm, "id"),
				getString(vm, "name"),
				getString(vm, "status"),
				ip,
				osName,
			}
		}

		output.PrintTable(rows, headers)

		// Show pagination info if available.
		output.PrintPageInfo(result, page, limit)
	},
}

// ---------- vm get ----------

var vmGetCmd = &cobra.Command{
	Use:   "get [VM_ID]",
	Short: "Get details of a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s", vmID)
		result := client.Get(path, nil, true)
		output.PrintItem(extractObj(result))
	},
}

// ---------- vm create ----------

var vmCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create a new virtual machine",
	Run: func(cmd *cobra.Command, args []string) {
		interactive, _ := cmd.Flags().GetBool("interactive")

		body := map[string]interface{}{}

		if interactive {
			body["name"] = PromptText("VM name:")
			body["hostname"] = PromptText("Hostname:")
			nodeID := SelectNode("Select node")
			body["node_id"] = nodeID
			body["os_template_id"] = SelectOSTemplate("Select OS template")
			body["plan_id"] = SelectPlan(nodeID, "Select plan")
			body["billing_type"] = SelectBillingType("Select billing type")
			sshKeys := SelectSSHKeys()
			if sshKeys != "" {
				body["ssh_keys"] = sshKeys
			}
			body["root_password"] = PromptPassword("Root password:")
		} else {
			name, _ := cmd.Flags().GetString("name")
			hostname, _ := cmd.Flags().GetString("hostname")
			nodeID, _ := cmd.Flags().GetString("node-id")
			osTemplateID, _ := cmd.Flags().GetString("os-template-id")
			rootPassword, _ := cmd.Flags().GetString("root-password")
			planID, _ := cmd.Flags().GetString("plan-id")
			billingType, _ := cmd.Flags().GetString("billing-type")
			cpu, _ := cmd.Flags().GetInt("cpu")
			ramMB, _ := cmd.Flags().GetInt("ram-mb")
			diskGB, _ := cmd.Flags().GetInt("disk-gb")
			sshKeys, _ := cmd.Flags().GetString("ssh-keys")

			if name == "" {
				fmt.Fprintln(os.Stderr, "Error: --name is required (or use -i for interactive mode)")
				os.Exit(1)
			}

			body["name"] = name
			if hostname != "" {
				body["hostname"] = hostname
			}
			if nodeID != "" {
				body["node_id"] = nodeID
			}
			if osTemplateID != "" {
				body["os_template_id"] = osTemplateID
			}
			if rootPassword != "" {
				body["root_password"] = rootPassword
			}
			if planID != "" {
				body["plan_id"] = planID
			}
			if billingType != "" {
				body["billing_type"] = billingType
			}
			if cpu > 0 {
				body["cpu"] = cpu
			}
			if ramMB > 0 {
				body["ram_mb"] = ramMB
			}
			if diskGB > 0 {
				body["disk_gb"] = diskGB
			}
			if sshKeys != "" {
				body["ssh_keys"] = sshKeys
			}
		}

		result := client.Post("/vms", body, true, 120)
		output.PrintSuccess("VM creation initiated.")
		output.PrintItem(extractObj(result))
	},
}

// ---------- vm delete ----------

var vmDeleteCmd = &cobra.Command{
	Use:   "delete [VM_ID]",
	Short: "Delete a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)

		if !Confirm(fmt.Sprintf("Are you sure you want to delete VM %s?", vmID)) {
			fmt.Println("Aborted.")
			return
		}

		path := fmt.Sprintf("/vms/%s", vmID)
		client.Delete(path, true)
		output.PrintSuccess(fmt.Sprintf("VM %s deleted.", vmID))
	},
}

// ---------- vm status ----------

var vmStatusCmd = &cobra.Command{
	Use:   "status [VM_ID]",
	Short: "Get status of a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/status", vmID)
		result := client.Get(path, nil, true)
		output.PrintItem(extractObj(result))
	},
}

// ---------- vm start ----------

var vmStartCmd = &cobra.Command{
	Use:   "start [VM_ID]",
	Short: "Start a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/start", vmID)
		client.Post(path, nil, true, 120)
		output.PrintSuccess(fmt.Sprintf("VM %s start initiated.", vmID))
	},
}

// ---------- vm stop ----------

var vmStopCmd = &cobra.Command{
	Use:   "stop [VM_ID]",
	Short: "Stop a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/stop", vmID)
		client.Post(path, nil, true, 120)
		output.PrintSuccess(fmt.Sprintf("VM %s stop initiated.", vmID))
	},
}

// ---------- vm reboot ----------

var vmRebootCmd = &cobra.Command{
	Use:   "reboot [VM_ID]",
	Short: "Reboot a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/reboot", vmID)
		client.Post(path, nil, true, 120)
		output.PrintSuccess(fmt.Sprintf("VM %s reboot initiated.", vmID))
	},
}

// ---------- vm rename ----------

var vmRenameCmd = &cobra.Command{
	Use:   "rename [VM_ID]",
	Short: "Rename a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)

		name, _ := cmd.Flags().GetString("name")
		if name == "" {
			name = PromptText("New VM name:")
		}

		path := fmt.Sprintf("/vms/%s/rename", vmID)
		body := map[string]interface{}{
			"name": name,
		}
		client.Patch(path, body, true)
		output.PrintSuccess(fmt.Sprintf("VM %s renamed to %s.", vmID, name))
	},
}

// ---------- vm reset-password ----------

var vmResetPasswordCmd = &cobra.Command{
	Use:   "reset-password [VM_ID]",
	Short: "Reset password for a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)

		username, _ := cmd.Flags().GetString("username")
		newPassword, _ := cmd.Flags().GetString("new-password")

		if username == "" {
			username = PromptText("Username:")
		}
		if newPassword == "" {
			newPassword = PromptPassword("New password:")
		}

		path := fmt.Sprintf("/vms/%s/reset-password", vmID)
		body := map[string]interface{}{
			"username":     username,
			"new_password": newPassword,
		}
		client.Post(path, body, true, 0)
		output.PrintSuccess(fmt.Sprintf("Password reset for user %s on VM %s.", username, vmID))
	},
}

// ---------- vm reset-network ----------

var vmResetNetworkCmd = &cobra.Command{
	Use:   "reset-network [VM_ID]",
	Short: "Reset network configuration for a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/reset-network", vmID)
		client.Post(path, nil, true, 0)
		output.PrintSuccess(fmt.Sprintf("Network reset initiated for VM %s.", vmID))
	},
}

// ---------- vm resize ----------

var vmResizeCmd = &cobra.Command{
	Use:   "resize [VM_ID]",
	Short: "Resize a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)

		planID, _ := cmd.Flags().GetString("plan-id")
		if planID == "" {
			planID = SelectPlan("", "Select new plan")
		}

		path := fmt.Sprintf("/vms/%s/resize", vmID)
		body := map[string]interface{}{
			"plan_id": planID,
		}
		client.Post(path, body, true, 0)
		output.PrintSuccess(fmt.Sprintf("VM %s resize initiated with plan %s.", vmID, planID))
	},
}

// ---------- vm reinstall ----------

var vmReinstallCmd = &cobra.Command{
	Use:   "reinstall [VM_ID]",
	Short: "Reinstall a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)

		if !Confirm(fmt.Sprintf("Are you sure you want to reinstall VM %s? This will destroy all data.", vmID)) {
			fmt.Println("Aborted.")
			return
		}

		osTemplateID, _ := cmd.Flags().GetString("os-template-id")
		rootPassword, _ := cmd.Flags().GetString("root-password")
		sshKeys, _ := cmd.Flags().GetString("ssh-keys")

		if osTemplateID == "" {
			osTemplateID = SelectOSTemplate("Select OS template")
		}
		if rootPassword == "" {
			rootPassword = PromptPassword("Root password:")
		}

		body := map[string]interface{}{
			"os_template_id": osTemplateID,
			"root_password":  rootPassword,
		}
		if sshKeys != "" {
			body["ssh_keys"] = sshKeys
		}

		path := fmt.Sprintf("/vms/%s/reinstall", vmID)
		client.Post(path, body, true, 0)
		output.PrintSuccess(fmt.Sprintf("VM %s reinstall initiated.", vmID))
	},
}

// ---------- vm console ----------

var vmConsoleCmd = &cobra.Command{
	Use:   "console [VM_ID]",
	Short: "Get console access URL for a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/console", vmID)
		result := client.Get(path, nil, true)
		obj := extractObj(result)

		// Try to display the console URL prominently.
		url := getString(obj, "url")
		if url == "" {
			url = getString(obj, "console_url")
		}
		if url != "" {
			output.PrintInfo(fmt.Sprintf("Console URL: %s", url))
		}
		output.PrintItem(obj)
	},
}

// ---------- vm network ----------

var vmNetworkCmd = &cobra.Command{
	Use:   "network [VM_ID]",
	Short: "Get network information for a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		path := fmt.Sprintf("/vms/%s/network", vmID)
		result := client.Get(path, nil, true)
		output.PrintItem(extractObj(result))
	},
}

// ---------- vm metrics ----------

var vmMetricsCmd = &cobra.Command{
	Use:   "metrics [VM_ID]",
	Short: "Get metrics for a virtual machine",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		vmID := resolveVMID(args)
		hours, _ := cmd.Flags().GetInt("hours")

		path := fmt.Sprintf("/vms/%s/metrics", vmID)
		params := map[string]string{
			"hours": strconv.Itoa(hours),
		}
		result := client.Get(path, params, true)
		obj := extractObj(result)

		// Format metrics output if structured data is available.
		formatted := false
		for _, key := range []string{"cpu", "memory", "disk", "network", "bandwidth"} {
			if val, ok := obj[key]; ok {
				fmt.Printf("\n%s:\n", strings.ToUpper(key))
				if m, ok := val.(map[string]interface{}); ok {
					for k, v := range m {
						fmt.Printf("  %-20s %v\n", k+":", v)
					}
					formatted = true
				} else {
					fmt.Printf("  %v\n", val)
					formatted = true
				}
			}
		}

		if !formatted {
			output.PrintItem(obj)
		}
	},
}
